Sources/ReferenceResolver.swift

```diff
@@ -53,22 +93,45 @@ class ReferenceResolver<T: ResolvableType> {
 
         /// Parse the pointer's path:
         let components = pointer.path.components(separatedBy: "/")
-        guard components.count == 3 && components[0] == "#" && components[1] == T.path else {
+        guard components.count == 4 && components[0] == "#" && components[2] == T.path else {
             throw ResolverError.invalidPath
         }
 
         /// Find the referenced builder (if any):
-        let name = components[2]
-        let referencedBuilder: T?
+        let name = components[3]
+        let referencedBuilder: Any?
         switch T.path {
-        case "definitions": referencedBuilder = swagger.definitionBuilders[name] as? T
-        case "parameters": referencedBuilder = swagger.parameterBuilders[name] as? T
-        case "responses": referencedBuilder = swagger.responseBuilders[name] as? T
+        case SchemaBuilder.path: referencedBuilder = swagger.components?.schemas[name]
+        case ParameterBuilder.path: referencedBuilder = swagger.components?.parameters[name]
+        case ResponseBuilder.path: referencedBuilder = swagger.components?.responses[name]
+        case SecuritySchemaBuilder.path: referencedBuilder = swagger.components?.securitySchemes[name]
+        case ExampleBuilder.path: referencedBuilder = swagger.components?.examples[name]
+        case RequestBodyBuilder.path: referencedBuilder = swagger.components?.requestBodies[name]
+        case HeaderBuilder.path: referencedBuilder = swagger.components?.headers[name]
+        case LinkBuilder.path: referencedBuilder = swagger.components?.headers[name]
+        case CallbackBuilder.path: referencedBuilder = swagger.components?.callbacks[name]
         default: throw ResolverError.unsupportedReference
         }
 
         // Check to see if the builder was found at the specified path:
-        guard let builder = referencedBuilder else {
+        guard let builderRef = (referencedBuilder as? Reference<T>) else {
+            throw ResolverError.unresolvedReference
+        }
+        
+        // Resolve references in reusable components itself
+        if case let Reference.pointer(pointer) = builderRef {
+            // Check to see if we have encountered a cyclic reference:
+            if context.resolvingReferences.contains(name) {
+                return Structure(name: name, structure: nil)
+            }
+            
+            // Push the resolving reference into the context to detect cyclic references.
+            context.resolvingReferences.insert(name)
+
+            return try resolve(swagger, pointer: pointer)
+        }
+        
+        guard case let Reference.value(builder) = builderRef else {
             throw ResolverError.unresolvedReference
         }
 
```

Comment:
### ReferenceResolver
- Path changed (e.g. '#/components/schemas/Tag' vs '#/shemas/Tag')
- Components object now contains Object or Reference, while previously reusable objects were not allowed to be references.

#### Merge proposal(s):
- Path could be handled by version check.
- Components now contains much more objects allowed for reuse and their type changed to Reference<T> instead of T. The referenced builders container also changed, thus possibly we could add protocol with method like `func referencedBuilder(for name: String) -> Any` or `Either<T, Reference<T>>` instead of `Any`, and support both versions in internal versioned if/else, still maintaining most of the code common.